tecken/base/symbolstorage.py

```diff
@@ -82,16 +82,22 @@ class SymbolStorage:
 
     """
 
-    def __init__(self, urls, try_url_index=None):
-        self.urls = urls
-        self.backends = [
-            StorageBucket(url, try_symbols=(i == try_url_index))
-            for i, url in enumerate(urls)
+    def __init__(
+        self, upload_url: str, download_urls: list[str], try_url: Optional[str] = None
+    ):
+        self.upload_backend = StorageBucket(upload_url)
+        self.download_backends = [
+            StorageBucket(url) for url in download_urls if url != upload_url
```

Comment:
It doesn't look like this distinction is used anywhere else -- and ultimately a few lines down, you merge these back together into `self.backends`. ~~Why separate them out in the parameters and here as separate intermediate variables?~~

Edit: I see now based on the next commit in the series: [Update possible_upload_urls() to use symbol_storage](https://github.com/mozilla-services/tecken/pull/2940/commits/5b199558d3b7a2346416e27eb647d2527578e5cd). These properties are accessed from outside the class.

Reply:
We need `upload_backend` separately to be able to use it for uploads. We need `try_backend` separately to be able to use it for try uploads, and append it to the backend list for try downloads. We don't really need `download_backends` separately, and it will be an empty list for our use case anyway, so I'll remove it from the instance.

Reply:
Tecken has the ability to scan multiple buckets when downloading files, and the URLs for downloading can be configured in the `SYMBOL_URLS` setting. When looking for a file, all URLs in that list are scanned, and if any of them contains the requested file it is used. You can see the code for that in `SymbolStorage._get()`. We don't make use of that ability, but it's there. You can see in `settings.py` that `SYMBOL_URL` uses `ListOf(str)`, which is everett's way of creating a list of strings.

When uploading files, Tecken only supports a single bucket. That bucket is configured in `UPLOAD_DEFAULT_BUCKET`. Since it's only a single url, it's a string, not a list of strings.

Because we use the same URL in both mentioned settings, we don't need to create the `StorageBucket` instance for that URL twice. For this reason I added `if url != upload_url` in the code above, so for us `download_backends` ends up being empty.

I could probably remove the ability to use multiple download buckets, but I don't want to fully rule out the option to use that during the migration, to make Tecken use the AWS bucket and the GCP bucket simultaneously. Currently that's not what the migration plan suggests to do, but it's possible we still hit some problem that makes that approach the most attractive option. And since the ability to use multiple download buckets is already there, I decided not to remove it for now.